src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -22,47 +31,132 @@ namespace Microsoft.CodeAnalysis.Editor.Implementation.Diagnostics
         where TTag : ITag
     {
-        private readonly object _uniqueKey = new object();
         private readonly IDiagnosticService _diagnosticService;
-        private readonly IForegroundNotificationService _notificationService;
-        private readonly IAsynchronousOperationListener _listener;
 
         protected AbstractDiagnosticsTaggerProvider(
             IDiagnosticService diagnosticService,
             IForegroundNotificationService notificationService,
             IAsynchronousOperationListener listener)
+            : base(listener, notificationService)
         {
             _diagnosticService = diagnosticService;
-            _notificationService = notificationService;
-            _listener = listener;
         }
 
-        protected internal abstract IEnumerable<Option<bool>> Options { get; }
+        protected override TaggerDelay AddedTagNotificationDelay => TaggerDelay.OnIdle;
+
+        protected override ITaggerEventSource CreateEventSource(ITextView textViewOpt, ITextBuffer subjectBuffer)
+        {
+            return TaggerEventSources.Compose(
+                TaggerEventSources.OnDocumentActiveContextChanged(subjectBuffer, TaggerDelay.Medium),
+                TaggerEventSources.OnWorkspaceRegistrationChanged(subjectBuffer, TaggerDelay.Medium),
+                TaggerEventSources.OnDiagnosticsChanged(subjectBuffer, _diagnosticService, TaggerDelay.Short));
+        }
+
         protected internal abstract bool IsEnabled { get; }
         protected internal abstract bool IncludeDiagnostic(DiagnosticData data);
         protected internal abstract ITagSpan<TTag> CreateTagSpan(bool isLiveUpdate, SnapshotSpan span, DiagnosticData data);
 
-        ITagger<T> ITaggerProvider.CreateTagger<T>(ITextBuffer buffer)
+        protected override Task ProduceTagsAsync(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag, int? caretPosition)
         {
-            return CreateTagger<T>(buffer);
+            ProduceTags(context, spanToTag);
+            return SpecializedTasks.EmptyTask;
         }
 
-        public IAccurateTagger<T> CreateTagger<T>(ITextBuffer buffer) where T : ITag
+        private void ProduceTags(TaggerContext<TTag> context, DocumentSnapshotSpan spanToTag)
         {
-            var tagger = buffer.Properties.GetOrCreateSingletonProperty(
-                _uniqueKey, () => new AggregatingTagger(this, buffer));
-            tagger.OnTaggerCreated();
-            return tagger as IAccurateTagger<T>;
+            if (!this.IsEnabled)
+            {
+                return;
+            }
+
+            var document = spanToTag.Document;
+            if (document == null)
+            {
+                return;
+            }
+
+            if (!document.TryGetText(out var sourceText))
+            {
+                return;
+            }
+
+            var editorSnapshot = sourceText.FindCorrespondingEditorTextSnapshot();
```

Comment:
How is this ever different than `spanToTag.SnapshotSpan.Snapshot`? This code and then the .TranslateTo in ProduceTags(...) implies with much vigor they can be different, but I don't see how.

Reply:
i think some of this logic dates back like 10ish years.  It's quite possible that it was back at a time where we had less invariants about Documents/SourceTexts/Snapshots.  I'm happy to remove (or assert it must be hte same).

Reply:
Well, it can only date 8 years ago... ;-)

Where did this move from originally? Was originally the snapshot you're mapping from related to the snapshot the diagnostics were computed for? It implies I can get stale spans, but since those buffers will be the same we're not correctly mapping things forward.

Reply:
This moved from: https://github.com/dotnet/roslyn/blob/master/src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs#L290

Reply:
Comment on that says:

```
                // Make sure we can find an editor snapshot for these errors.  Otherwise we won't
                // be able to make ITagSpans for them.  If we can't, just bail out.  This happens
                // when the solution crawler is very far behind.  However, it will have a more
                // up to date document within it that it will eventually process.  Until then
                // we just keep around the stale tags we have.
```

I think a lot of the complexity here was that in the old system you had notifications from the diagnostic engine about a set of data, and we were trying to relate it to the editor-based data.  There were moving at different speeds, with different invariants, and we had code like this to try to be resilient to that sort of thing.

If we're now just querying the diagnostic subsystem with teh correct tagger info we have, i don't think we have any issues adn we can likely clean this up even more.

Reply:
Yup, looking at this, i don't think we need it anymore.

Reply:
old one did translate to since DiagnosticUpdated event is not part of tag event source. due to that it didn't update internal interval tree when diagnostic is updated, rather it just hold onto snapshot at the moment. and later did "translate to" move spans to right one itself.

this new one no longer does that, now DiagnosticUpdated is part of tag event source. when diagnostic changed, it save spans in taggers internal interval tree and let tagger's own mechanism to handle translate to.

Reply:
"// Make sure we can find an editor snapshot for these errors.  Otherwise we won't
                // be able to make ITagSpans for them.  If we can't, just bail out.  This happens
                // when the solution crawler is very far behind.  However, it will have a more
                // up to date document within it that it will eventually process.  Until then
                // we just keep around the stale tags we have."

agree this is not needed anymore. since there is no longer 2 documents as we used to.